apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/pac/common/ClaimDao.java

```diff
@@ -133,6 +129,82 @@ public <T> List<T> findAllByMbiAttribute(
+   * @param isMbiSearchValueHashed True if the mbiSearchValue is a hashed MBI.
+   * @param lastUpdated The range of lastUpdated values to search on.
+   * @param serviceDate The range of the desired service date to search on.
+   * @return a {@link Predicate} to be used in query where clause
+   * @param <T> The entity type being retrieved.
+   */
+  @VisibleForTesting
+  <T> List<Predicate> createStandardPredicatesForMbiLookup(
+      CriteriaBuilder builder,
+      Root<T> root,
+      ResourceTypeV2<?, T> resourceType,
+      String mbiSearchValue,
+      boolean isMbiSearchValueHashed,
+      DateRangeParam lastUpdated,
+      DateRangeParam serviceDate) {
+    final String mbiRecordAttributeName = resourceType.getEntityMbiRecordAttribute();
+    final List<String> serviceDateAttributeNames = resourceType.getEntityServiceDateAttributes();
+    final List<Predicate> predicates = new ArrayList<>();
+    final Path<Object> mbiRecord = root.get(mbiRecordAttributeName);
+    predicates.add(createMbiPredicate(mbiRecord, mbiSearchValue, isMbiSearchValueHashed, builder));
+    if (isDateRangePresent(lastUpdated)) {
+      predicates.add(lastUpdatedPredicate(root, lastUpdated, builder));
+    }
+    if (isDateRangePresent(serviceDate) && !serviceDateAttributeNames.isEmpty()) {
+      final var serviceDatePredicate =
+          createServiceDatePredicates(builder, root, serviceDate, serviceDateAttributeNames);
+      predicates.add(serviceDatePredicate);
+    }
+    return predicates;
+  }
+
+  /**
+   * Shorthand for checking that a date range parameter has been populated by our caller.
+   *
+   * @param dateRangeParam the param to check
+   * @return true if the param contains a date range
+   */
+  private boolean isDateRangePresent(DateRangeParam dateRangeParam) {
+    return dateRangeParam != null && !dateRangeParam.isEmpty();
+  }
+
+  /**
+   * Creates a service date predicate for each attribute name in the list and combines them using
+   * logical OR.
+   *
+   * @param builder The {@link CriteriaBuilder} being used in current query.
+   * @param root The {@link Root} for the claim in the query
+   * @param serviceDate The range of the desired service date to search on.
+   * @param serviceDateAttributeNames List of service date attribute names.
+   * @return a {@link Predicate} to be used in query where clause
+   * @param <T> The entity type being retrieved.
+   */
+  @VisibleForTesting
+  <T> Predicate createServiceDatePredicates(
+      CriteriaBuilder builder,
+      Root<T> root,
+      DateRangeParam serviceDate,
+      List<String> serviceDateAttributeNames) {
+    final var serviceDatePredicates =
+        serviceDateAttributeNames.stream()
+            .map(root::<LocalDate>get)
+            .map(dateColumn -> serviceDatePredicate(builder, serviceDate, dateColumn))
+            .toArray(Predicate[]::new);
+    return builder.or(serviceDatePredicates);
```

Comment:
The first map calls root.get(x) to get a Path for the date attribute and the serviceDatePredicate converts that into a Predicate.  The toArray just combines them all into an array of predicates.  Using multiple maps is generally clearer since each lambda is trivial rather than having a multi-line lambda that does both things.   The only other alternative would be a loop and I find those to be noisier and harder to follow than simple chained calls.

Are the method references confusing?  IDEA indicates they are preferred to simple lambdas so that's why I used them.  Personally I find simple lambdas clearer.

Reply:
I think it's a mix between the method names, chained maps, and the method reference syntax. I would generally agree with IntelliJ that the method ref syntax is nicer, but when generics get involved and the object name is not super obvious what it is (like "root") it becomes a little more confusing. Also yeah at least to me .toArray(Predicate[]::new); is kinda hard to parse quickly.

Maybe just nix the shorthand on the first map and the toArray and it might read a little nicer without devolving it all into loops.

Reply:
I agree the JPA criteria query classes are pretty hard to follow.  Is there a naming convention on the team for how to name CriteriaQuery, Path, and Root objects?

Reply:
I have replaced the method refs with lambdas and given their parameters meaninful names.  Generally I'm not a fan of method refs anyway.  I've changed my IDEA setting to mark those opportunities as weak warnings instead of warnings to shut it up. :-)